packages/medusa/src/models/batch-job.ts

```diff
@@ -1,30 +1,85 @@
-import { BeforeInsert, Column, Entity, PrimaryColumn } from "typeorm"
+import { AfterLoad, BeforeInsert, Column, Entity, JoinColumn, ManyToOne } from "typeorm"
 import { BatchJobStatus } from "../types/batch-job"
-import { DbAwareColumn } from "../utils/db-aware-column"
+import { DbAwareColumn, resolveDbType } from "../utils/db-aware-column"
 import { SoftDeletableEntity } from "../interfaces/models/soft-deletable-entity"
 import { generateEntityId } from "../utils/generate-entity-id"
+import { User } from "./user"
 
 @Entity()
 export class BatchJob extends SoftDeletableEntity {
   @DbAwareColumn({ type: "text" })
   type: string
 
-  @DbAwareColumn({ type: "enum", enum: BatchJobStatus })
-  status: BatchJobStatus
-
-  @Column({ type: "text", nullable: true })
+  @Column({ nullable: true })
   created_by: string | null
 
+  @ManyToOne(() => User)
+  @JoinColumn({ name: "created_by" })
+  created_by_user: User
+
   @DbAwareColumn({ type: "jsonb", nullable: true })
-  context: Record<string, unknown>
+  context: { retry_count?: number; max_retry?: number } & Record<string, unknown>
 
   @DbAwareColumn({ type: "jsonb", nullable: true })
   result: Record<string, unknown>
 
+  @Column({ type: "boolean", nullable: false, default: false })
+  dry_run: boolean = false;
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  pre_processing_at?: Date
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  processing_at?: Date
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  completed_at?: Date
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  canceled_at?: Date
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  ready_at?: Date
+
+  @Column({ type: resolveDbType("timestamptz"), nullable: true })
+  failed_at?: Date
+
+  status: BatchJobStatus
+
+  @AfterLoad()
+  loadStatus(): void {
+    /* Always keep the status order consistent. */
+    if (this.pre_processing_at) {
+      this.status = BatchJobStatus.PRE_PROCESSING
+    }
+    if (this.ready_at) {
+      this.status = BatchJobStatus.READY
+    }
+    if (this.processing_at) {
+      this.status = BatchJobStatus.PROCESSING
+    }
+    if (this.completed_at) {
+      this.status = BatchJobStatus.COMPLETED
+    }
+    if (this.canceled_at) {
+      this.status = BatchJobStatus.CANCELED
+    }
+    if (this.failed_at) {
+      this.status = BatchJobStatus.FAILED
+    }
+
+    this.status = this.status ?? BatchJobStatus.CREATED
+  }
+
   @BeforeInsert()
   private beforeInsert(): void {
     this.id = generateEntityId(this.id, "batch")
   }
+
+  toJSON() {
+    this.loadStatus()
+    return this
+  }
```

Comment:
This is the default approach for a toJSON method for an object. This is the one that is called when an object is sent though the network and allow to prepare an object before it is stringify. All object have a default toJSON implementation that return an object

Reply:
alright but this doesn't remove the methods on the object?

```
const cleanBatch = retrievedBatchJob.toJSON()

assert("beforeInsert" in batch === false) // this would fail right?
```

Reply:
This is also non-blocking but just a thought

Reply:
Yes it is still on the object unless we hide it if I use the approach of Object.defineProperty and set enumerable/configurable to false. Weyt?